backend/services/implementations/testSessionService.ts

```diff
@@ -334,10 +334,48 @@ class TestSessionService implements ITestSessionService {
   async updateTestSession(
     id: string,
     testSession: TestSessionRequestDTO,
+    nowDate?: Date,
   ): Promise<TestSessionResponseDTO> {
     let updatedTestSession: TestSession | null;
 
     try {
+      const date = nowDate ?? new Date();
+      const oldTestSession: TestSession | null = await MgTestSession.findById(
+        id,
+      );
+      if (!oldTestSession) {
+        throw new Error(`Test Session id ${id} not found`);
+      }
+
+      // If the test session is past, we don't allow teachers to update the test session at all
+      if (oldTestSession.endDate < date) {
+        throw new Error(
+          `Test Session id ${id} has already ended and so cannot be updated`,
+        );
+      }
+
+      // If the test session is active, we don't allow teachers to update the assessment, class, or start date
+      const isActive = oldTestSession.startDate <= date;
+      const keysToValidate = ["test", "class", "startDate"];
```

Comment:
Oh sorry, can we do an exclusive check instead? Like a `validKeys` set and we have an error if there's a modification not in the valid keys? That way we get the benefit from my previous comment about data model updates

Reply:
hopefully i addressed it this time.

also side thought jw, should we be passing a partial objects on update (only fields that user intends to edit) or the whole request object

Reply:
Agree, partial object would be ideal. But we can keep as is for now I think, nbd